File: src/animate/easing.ts
/** Maps normalised time to normalised progress. Both are 0–1, though some curves overshoot. */
export type EasingFn = (t: number) => number

/** An easing named from the catalogue, or supplied directly. */
export type Easing = EasingName | EasingFn

const clamp01 = (t: number): number => (t < 0 ? 0 : t > 1 ? 1 : t)

/** Turns an ease-in curve into its out and in-out counterparts, so each family is written once. */
const family = (easeIn: EasingFn) => ({
  in: (t: number) => easeIn(t),
  out: (t: number) => 1 - easeIn(1 - t),
  inOut: (t: number) => (t < 0.5 ? easeIn(t * 2) / 2 : 1 - easeIn((1 - t) * 2) / 2),
})

/** Overshoot for the back family, the constant CSS and every animation library settled on. */
const BACK_OVERSHOOT = 1.70158
/** Period of the elastic oscillation, as a fraction of the duration. */
const ELASTIC_PERIOD = 0.3
const BOUNCE_N = 7.5625
const BOUNCE_D = 2.75

const quad = family(t => t * t)
const cubic = family(t => t * t * t)
const quart = family(t => t * t * t * t)
const quint = family(t => t * t * t * t * t)
const sine = family(t => 1 - Math.cos((t * Math.PI) / 2))
const expo = family(t => (t === 0 ? 0 : Math.pow(2, 10 * t - 10)))
const circ = family(t => 1 - Math.sqrt(1 - t * t))
const back = family(t => (BACK_OVERSHOOT + 1) * t * t * t - BACK_OVERSHOOT * t * t)
const elastic = family(t => {
  if (t === 0 || t === 1) return t
  return -Math.pow(2, 10 * t - 10) * Math.sin(((t * 10 - 10.75) * (2 * Math.PI)) / ELASTIC_PERIOD)
})

/** Bounce is defined by its out form — the piecewise decay everyone recognises — and mirrored. */
const outBounce: EasingFn = t => {
  if (t < 1 / BOUNCE_D) return BOUNCE_N * t * t
  if (t < 2 / BOUNCE_D) return BOUNCE_N * (t -= 1.5 / BOUNCE_D) * t + 0.75
  if (t < 2.5 / BOUNCE_D) return BOUNCE_N * (t -= 2.25 / BOUNCE_D) * t + 0.9375
  return BOUNCE_N * (t -= 2.625 / BOUNCE_D) * t + 0.984375
}
const bounce = family(t => 1 - outBounce(1 - t))

/**
 * The standard easing catalogue.
 *
 * Every entry clamps its input, so a track that runs past its own duration holds at its end value
 * rather than continuing to accelerate off the curve.
 * @example
 * ```ts
 * easings.outCubic(0.5)  // 0.875
 * track({ from: 0, to: 1, duration: 1, ease: 'outCubic' })
 * ```
 */
export const easings = {
  /** No easing: the value moves at a constant rate from start to finish. */
  linear: (t: number) => clamp01(t),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by t squared — the gentlest of the power curves.
   */
  inQuad: (t: number) => quad.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by t squared — the gentlest of the power curves.
   */
  outQuad: (t: number) => quad.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by t squared — the gentlest of the power curves.
   */
  inOutQuad: (t: number) => quad.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by t cubed — the usual choice when nothing else is called for.
   */
  inCubic: (t: number) => cubic.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by t cubed — the usual choice when nothing else is called for.
   */
  outCubic: (t: number) => cubic.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by t cubed — the usual choice when nothing else is called for.
   */
  inOutCubic: (t: number) => cubic.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by t to the fourth — noticeably sharper than cubic.
   */
  inQuart: (t: number) => quart.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by t to the fourth — noticeably sharper than cubic.
   */
  outQuart: (t: number) => quart.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by t to the fourth — noticeably sharper than cubic.
   */
  inOutQuart: (t: number) => quart.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by t to the fifth — the steepest of the plain powers.
   */
  inQuint: (t: number) => quint.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by t to the fifth — the steepest of the plain powers.
   */
  outQuint: (t: number) => quint.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by t to the fifth — the steepest of the plain powers.
   */
  inOutQuint: (t: number) => quint.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a quarter turn of a sine wave — the softest curve here, barely a curve at all.
   */
  inSine: (t: number) => sine.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a quarter turn of a sine wave — the softest curve here, barely a curve at all.
   */
  outSine: (t: number) => sine.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a quarter turn of a sine wave — the softest curve here, barely a curve at all.
   */
  inOutSine: (t: number) => sine.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a power of two — almost flat, then a rush — the most extreme of these.
   */
  inExpo: (t: number) => expo.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a power of two — almost flat, then a rush — the most extreme of these.
   */
  outExpo: (t: number) => expo.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a power of two — almost flat, then a rush — the most extreme of these.
   */
  inOutExpo: (t: number) => expo.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a quarter circle — idles, then leaves abruptly.
   */
  inCirc: (t: number) => circ.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a quarter circle — idles, then leaves abruptly.
   */
  outCirc: (t: number) => circ.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a quarter circle — idles, then leaves abruptly.
   */
  inOutCirc: (t: number) => circ.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a cubic that overshoots — pulls back before it sets off, and overshoots on arrival.
   */
  inBack: (t: number) => back.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a cubic that overshoots — pulls back before it sets off, and overshoots on arrival.
   */
  outBack: (t: number) => back.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a cubic that overshoots — pulls back before it sets off, and overshoots on arrival.
   */
  inOutBack: (t: number) => back.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a decaying sine — overshoots and wobbles to a stop, like something sprung.
   */
  inElastic: (t: number) => elastic.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a decaying sine — overshoots and wobbles to a stop, like something sprung.
   */
  outElastic: (t: number) => elastic.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a decaying sine — overshoots and wobbles to a stop, like something sprung.
   */
  inOutElastic: (t: number) => elastic.inOut(clamp01(t)),

  /**
   * Starts at rest and accelerates, so the movement lands at full speed.
   *
   * Shaped by a series of parabolas — lands and bounces, each one smaller than the last.
   */
  inBounce: (t: number) => bounce.in(clamp01(t)),

  /**
   * Starts at full speed and slows to rest, which is what most interface motion wants.
   *
   * Shaped by a series of parabolas — lands and bounces, each one smaller than the last.
   */
  outBounce: (t: number) => bounce.out(clamp01(t)),

  /**
   * Accelerates from rest and slows to rest again, symmetric about the midpoint.
   *
   * Shaped by a series of parabolas — lands and bounces, each one smaller than the last.
   */
  inOutBounce: (t: number) => bounce.inOut(clamp01(t)),
} satisfies Record<string, EasingFn>

/** The name of any built-in easing — every key of {@link easings}. */
export type EasingName = keyof typeof easings

/** How close the solver has to get before it accepts a value of t. */
const BEZIER_EPSILON = 1e-6
const BEZIER_MAX_ITERATIONS = 12

/**
 * Builds the CSS `cubic-bezier(x1, y1, x2, y2)` curve.
 *
 * The curve is parametric, so drawing it at a given time means first finding the parameter whose x
 * equals that time. Newton's method converges in a few steps for well-behaved curves; a bisection
 * fallback covers the steep ones, where the derivative approaches zero and Newton stalls.
 * @example
 * ```ts
 * const easeInOut = cubicBezier(0.42, 0, 0.58, 1) // the CSS ease-in-out curve
 * ```
 */
export function cubicBezier(x1: number, y1: number, x2: number, y2: number): EasingFn {
  const curve = (a: number, b: number, t: number) => {
    const c = 3 * a
    const bTerm = 3 * (b - a) - c
    const aTerm = 1 - c - bTerm
    return ((aTerm * t + bTerm) * t + c) * t
  }
  const slope = (a: number, b: number, t: number) => {
    const c = 3 * a
    const bTerm = 3 * (b - a) - c
    const aTerm = 1 - c - bTerm
    return (3 * aTerm * t + 2 * bTerm) * t + c
  }

  return (time: number): number => {
    const target = clamp01(time)
    if (target === 0 || target === 1) return target

    let t = target
    for (let i = 0; i < BEZIER_MAX_ITERATIONS; i++) {
      const x = curve(x1, x2, t) - target
      if (Math.abs(x) < BEZIER_EPSILON) return curve(y1, y2, t)
      const d = slope(x1, x2, t)
      if (Math.abs(d) < BEZIER_EPSILON) break
      t -= x / d
    }

    // Newton stalled on a near-flat section; bisect, which cannot.
    let low = 0
    let high = 1
    t = target
    while (high - low > BEZIER_EPSILON) {
      if (curve(x1, x2, t) < target) low = t
      else high = t
      t = (low + high) / 2
    }
    return curve(y1, y2, t)
  }
}

/**
 * Quantises progress into `count` equal jumps, as CSS `steps(count, end)` does.
 *
 * Floors rather than rounds, which is what makes it hold each step for its full width: a value a
 * hair below the next boundary belongs to the step it is still inside. The final instant is pinned
 * to 1 so the animation lands on its end value rather than a step short of it.
 */
export function steps(count: number): EasingFn {
  if (!Number.isInteger(count) || count < 1) {
    throw new Error(`[canvas] steps() needs at least 1 step (got ${count})`)
  }
  return (t: number) => {
    const clamped = clamp01(t)
    return clamped >= 1 ? 1 : Math.floor(clamped * count) / count
  }
}

/** Resolves a name or function into an easing, defaulting to linear. */
export function resolveEasing(easing: Easing | undefined): EasingFn {
  if (easing === undefined) return easings.linear
  if (typeof easing === 'function') return easing

  const found = easings[easing]
  if (!found) {
    throw new Error(`[canvas] "${easing}" is not a known easing — see \`easings\` for the catalogue`)
  }
  return found
}

File: crates/meo-canvas-core/src/animate/easing.rs
//! Timing curves: the catalogue, `cubic-bezier` and `steps`.
//!
//! In `f64` so the two surfaces compare exactly -- JavaScript has only `f64` --
//! narrowing to `f32` once, when the value reaches a style. The constants are
//! CSS's standard ones, checked against vectors recorded from v9; a curve
//! rewritten from its name differs in the fourth decimal.

//! # Why this module refuses `mul_add`
//!
//! A fused multiply-add rounds once where `a * b + c` rounds twice, so its last
//! bit differs from JavaScript's, which has no fused form. This module is
//! compared with `==` on `f64`, so the last bit is the whole assertion.
//!
//! The rule follows the comparison, not the architecture: fused arithmetic is
//! fine where a tolerance dwarfs it -- `chrome_blend.rs` allows 1 in 255
//! against a `1e-16` difference, and `paint.rs` likewise -- and becomes wrong
//! the moment such a comparison is made exact.

#![expect(
    clippy::float_cmp,
    clippy::while_float,
    clippy::manual_midpoint,
    reason = "v9's arithmetic, kept in v9's shape. The equality tests are \
              exact by design -- `t == 0.0` and `t == 1.0` are the endpoints a \
              curve is pinned to, not approximations of them -- and the \
              bisection loop's `high - low > EPSILON` and `(low + high) / 2.0` \
              are the solver v9 runs. `f64::midpoint` computes a midpoint by a \
              different route and may round elsewhere, which is a difference \
              this module is compared bit-for-bit on."
)]
#![expect(
    clippy::suboptimal_flops,
    reason = "the lint wants a fused multiply-add, which rounds once where \
              JavaScript rounds twice. This module is implemented twice and \
              compared bit-for-bit against `tests/assets/animate/*.tsv`, so \
              agreement is the objective and accuracy is not. `expect` rather \
              than `allow`: if these expressions ever stop tripping the lint, \
              this stops being load-bearing and should be removed."
)]

use crate::Error;

/// Clamped, because a track running past its own duration should hold at its
/// end value rather than accelerate off the curve.
const fn clamp01(t: f64) -> f64 {
    t.clamp(0.0, 1.0)
}

/// The overshoot the `back` family pulls back by, which is CSS's own constant.
const BACK_OVERSHOOT: f64 = 1.701_58;
/// The elastic oscillation's period, as a fraction of the duration.
const ELASTIC_PERIOD: f64 = 0.3;
/// The bounce parabola's height and the divisions of its four landings.
const BOUNCE_N: f64 = 7.562_5;
/// How the bounce's timeline divides: the four landings are fractions of this.
const BOUNCE_D: f64 = 2.75;

/// The `out` counterpart of an `in` curve: the same shape, reversed twice.
fn out_of(ease_in: impl Fn(f64) -> f64, t: f64) -> f64 {
    1.0 - ease_in(1.0 - t)
}

/// The `inOut` counterpart: the `in` curve to the midpoint, mirrored after it.
fn in_out_of(ease_in: impl Fn(f64) -> f64 + Copy, t: f64) -> f64 {
    if t < 0.5 {
        ease_in(t * 2.0) / 2.0
    } else {
        1.0 - ease_in((1.0 - t) * 2.0) / 2.0
    }
}

fn quad(t: f64) -> f64 {
    t * t
}

fn cubic(t: f64) -> f64 {
    t * t * t
}

fn quart(t: f64) -> f64 {
    t * t * t * t
}

fn quint(t: f64) -> f64 {
    t * t * t * t * t
}

fn sine(t: f64) -> f64 {
    1.0 - (t * std::f64::consts::PI / 2.0).cos()
}

/// Zero is a special case rather than an approximation: `2^-10` is `0.000977`,
/// which is a visible offset at the start of a curve that should begin at rest.
fn expo(t: f64) -> f64 {
    if t == 0.0 {
        0.0
    } else {
        (10.0 * t - 10.0).exp2()
    }
}

fn circ(t: f64) -> f64 {
    1.0 - (1.0 - t * t).sqrt()
}

fn back(t: f64) -> f64 {
    (BACK_OVERSHOOT + 1.0) * t * t * t - BACK_OVERSHOOT * t * t
}

/// Both ends are exact rather than computed: the oscillation's envelope is
/// zero at neither end by arithmetic, and a curve that starts at `-0.0` or
/// lands a hair off 1 is one that does not settle.
fn elastic(t: f64) -> f64 {
    if t == 0.0 || t == 1.0 {
        return t;
    }
    -(10.0 * t - 10.0).exp2()
        * ((t * 10.0 - 10.75) * (2.0 * std::f64::consts::PI) / ELASTIC_PERIOD)
            .sin()
}

/// The bounce is written in its `out` form -- the piecewise decay everyone
/// recognises -- and the other two are mirrors of it.
fn out_bounce(t: f64) -> f64 {
    if t < 1.0 / BOUNCE_D {
        return BOUNCE_N * t * t;
    }
    if t < 2.0 / BOUNCE_D {
        let t = t - 1.5 / BOUNCE_D;
        return BOUNCE_N * t * t + 0.75;
    }
    if t < 2.5 / BOUNCE_D {
        let t = t - 2.25 / BOUNCE_D;
        return BOUNCE_N * t * t + 0.937_5;
    }
    let t = t - 2.625 / BOUNCE_D;
    BOUNCE_N * t * t + 0.984_375
}

fn bounce(t: f64) -> f64 {
    1.0 - out_bounce(1.0 - t)
}

/// How close the solver gets to a time before accepting the parameter: `1e-6`
/// in `x`, the most the algorithm promises, and so the tolerance for any
/// cross-surface comparison of [`cubic_bezier`].
pub const BEZIER_EPSILON: f64 = 1e-6;
/// Newton's iterations before the bisection fallback takes over.
const BEZIER_MAX_ITERATIONS: usize = 12;

/// The CSS `cubic-bezier(x1, y1, x2, y2)` curve.
///
/// Drawing it at a time means finding the parameter whose x is that time:
/// Newton, falling back to bisection where a near-flat section stalls it or
/// twelve iterations pass.
#[must_use]
pub const fn cubic_bezier(x1: f64, y1: f64, x2: f64, y2: f64) -> CubicBezier {
    CubicBezier { x1, y1, x2, y2 }
}

/// A `cubic-bezier` curve, as a value a caller can hold.
///
/// A struct field, a return type or an element of a `Vec`, which `impl Fn`
/// cannot be. [`CubicBezier::at`] matches [`Easing::at`]; for a closure,
/// `move |t| curve.at(t)`.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct CubicBezier {
    x1: f64,
    y1: f64,
    x2: f64,
    y2: f64,
}

impl CubicBezier {
    /// The curve's value at a normalised time.
    #[must_use]
    pub fn at(&self, time: f64) -> f64 {
        let (x1, y1, x2, y2) = (self.x1, self.y1, self.x2, self.y2);
        let target = clamp01(time);
        if target == 0.0 || target == 1.0 {
            return target;
        }

        let mut t = target;
        for _ in 0..BEZIER_MAX_ITERATIONS {
            let x = bezier(x1, x2, t) - target;
            if x.abs() < BEZIER_EPSILON {
                return bezier(y1, y2, t);
            }
            let slope = bezier_slope(x1, x2, t);
            if slope.abs() < BEZIER_EPSILON {
                break;
            }
            t -= x / slope;
        }

        let (mut low, mut high) = (0.0, 1.0);
        t = target;
        while high - low > BEZIER_EPSILON {
            if bezier(x1, x2, t) < target {
                low = t;
            } else {
                high = t;
            }
            t = (low + high) / 2.0;
        }
        bezier(y1, y2, t)
    }
}

/// One axis of the curve at parameter `t`, in Horner form.
fn bezier(a: f64, b: f64, t: f64) -> f64 {
    let c = 3.0 * a;
    let b_term = 3.0 * (b - a) - c;
    let a_term = 1.0 - c - b_term;
    ((a_term * t + b_term) * t + c) * t
}

/// That axis's derivative, which is what Newton steps along.
fn bezier_slope(a: f64, b: f64, t: f64) -> f64 {
    let c = 3.0 * a;
    let b_term = 3.0 * (b - a) - c;
    let a_term = 1.0 - c - b_term;
    (3.0 * a_term * t + 2.0 * b_term) * t + c
}

/// How many steps a [`steps`] curve may be built with.
///
/// One is the floor: a zero-step curve has no width to hold a value for.
pub const MIN_STEPS: u32 = 1;

/// Quantises progress into `count` equal jumps, as CSS `steps(count, end)`.
/// Floors, so each step holds its full width, and pins the final instant to 1
/// so the animation lands on its end value.
///
/// # Errors
///
/// Returns [`Error::Steps`] for a count below [`MIN_STEPS`].
pub const fn steps(count: u32) -> Result<Steps, Error> {
    if count < MIN_STEPS {
        return Err(Error::Steps(count));
    }
    Ok(Steps { count })
}

/// A `steps()` curve as a value a caller can hold, named for the reason
/// [`CubicBezier`] is. Built only through [`steps`], so a `Steps` holds at
/// least [`MIN_STEPS`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Steps {
    count: u32,
}

impl Steps {
    /// The curve's value at a normalised time.
    #[must_use]
    pub fn at(&self, t: f64) -> f64 {
        let count = f64::from(self.count);
        let clamped = clamp01(t);
        if clamped >= 1.0 {
            1.0
        } else {
            (clamped * count).floor() / count
        }
    }
}

/// The standard easing catalogue, by name. Every entry clamps its input, so a
/// track past its duration holds at its end value; `Back` and `Elastic` still
/// overshoot within that range.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[expect(
    missing_docs,
    reason = "thirty variants whose names are their definitions: `OutCubic` \
              is the cubic curve's out form, and a line saying so on each \
              would be thirty lines nobody reads. The families are documented \
              where they are implemented."
)]
pub enum Easing {
    /// No easing: a constant rate from start to finish.
    Linear,
    InQuad,
    OutQuad,
    InOutQuad,
    InCubic,
    OutCubic,
    InOutCubic,
    InQuart,
    OutQuart,
    InOutQuart,
    InQuint,
    OutQuint,
    InOutQuint,
    InSine,
    OutSine,
    InOutSine,
    InExpo,
    OutExpo,
    InOutExpo,
    InCirc,
    OutCirc,
    InOutCirc,
    InBack,
    OutBack,
    InOutBack,
    InElastic,
    OutElastic,
    InOutElastic,
    InBounce,
    OutBounce,
    InOutBounce,
}

impl Easing {
    /// Every curve, so a test can walk the catalogue rather than list it.
    pub const ALL: [Self; 31] = [
        Self::Linear,
        Self::InQuad,
        Self::OutQuad,
        Self::InOutQuad,
        Self::InCubic,
        Self::OutCubic,
        Self::InOutCubic,
        Self::InQuart,
        Self::OutQuart,
        Self::InOutQuart,
        Self::InQuint,
        Self::OutQuint,
        Self::InOutQuint,
        Self::InSine,
        Self::OutSine,
        Self::InOutSine,
        Self::InExpo,
        Self::OutExpo,
        Self::InOutExpo,
        Self::InCirc,
        Self::OutCirc,
        Self::InOutCirc,
        Self::InBack,
        Self::OutBack,
        Self::InOutBack,
        Self::InElastic,
        Self::OutElastic,
        Self::InOutElastic,
        Self::InBounce,
        Self::OutBounce,
        Self::InOutBounce,
    ];

    /// The curve's value at a normalised time.
    #[must_use]
    pub fn at(self, time: f64) -> f64 {
        let t = clamp01(time);
        match self {
            Self::Linear => t,
            Self::InQuad => quad(t),
            Self::OutQuad => out_of(quad, t),
            Self::InOutQuad => in_out_of(quad, t),
            Self::InCubic => cubic(t),
            Self::OutCubic => out_of(cubic, t),
            Self::InOutCubic => in_out_of(cubic, t),
            Self::InQuart => quart(t),
            Self::OutQuart => out_of(quart, t),
            Self::InOutQuart => in_out_of(quart, t),
            Self::InQuint => quint(t),
            Self::OutQuint => out_of(quint, t),
            Self::InOutQuint => in_out_of(quint, t),
            Self::InSine => sine(t),
            Self::OutSine => out_of(sine, t),
            Self::InOutSine => in_out_of(sine, t),
            Self::InExpo => expo(t),
            Self::OutExpo => out_of(expo, t),
            Self::InOutExpo => in_out_of(expo, t),
            Self::InCirc => circ(t),
            Self::OutCirc => out_of(circ, t),
            Self::InOutCirc => in_out_of(circ, t),
            Self::InBack => back(t),
            Self::OutBack => out_of(back, t),
            Self::InOutBack => in_out_of(back, t),
            Self::InElastic => elastic(t),
            Self::OutElastic => out_of(elastic, t),
            Self::InOutElastic => in_out_of(elastic, t),
            Self::InBounce => bounce(t),
            Self::OutBounce => out_of(bounce, t),
            Self::InOutBounce => in_out_of(bounce, t),
        }
    }

    /// The name v9 spells this curve with, which is the name the vector table
    /// and the TypeScript surface both use.
    #[must_use]
    pub const fn name(self) -> &'static str {
        match self {
            Self::Linear => "linear",
            Self::InQuad => "inQuad",
            Self::OutQuad => "outQuad",
            Self::InOutQuad => "inOutQuad",
            Self::InCubic => "inCubic",
            Self::OutCubic => "outCubic",
            Self::InOutCubic => "inOutCubic",
            Self::InQuart => "inQuart",
            Self::OutQuart => "outQuart",
            Self::InOutQuart => "inOutQuart",
            Self::InQuint => "inQuint",
            Self::OutQuint => "outQuint",
            Self::InOutQuint => "inOutQuint",
            Self::InSine => "inSine",
            Self::OutSine => "outSine",
            Self::InOutSine => "inOutSine",
            Self::InExpo => "inExpo",
            Self::OutExpo => "outExpo",
            Self::InOutExpo => "inOutExpo",
            Self::InCirc => "inCirc",
            Self::OutCirc => "outCirc",
            Self::InOutCirc => "inOutCirc",
            Self::InBack => "inBack",
            Self::OutBack => "outBack",
            Self::InOutBack => "inOutBack",
            Self::InElastic => "inElastic",
            Self::OutElastic => "outElastic",
            Self::InOutElastic => "inOutElastic",
            Self::InBounce => "inBounce",
            Self::OutBounce => "outBounce",
            Self::InOutBounce => "inOutBounce",
        }
    }
}
